In [ ]:
import pandas as pd

Loading Dataset

In [ ]:
df = pd.read_csv('/content/synthetic_fnb_order_items_bridge_1000.csv')
df.head()

,order_id,product_name,order_date,order_type,payment_method,rating,price,region,preparation_time_minutes,quantity,line_total,employee_id
0,ORD00334,Rice Bowl,2026-02-27 08:33,Takeaway,Card,5.0,36.76,South,33,4,147.04,EMP026
1,ORD00284,Noodles,2026-03-03 14:35,Takeaway,Cash,3.0,14.31,East,32,3,42.93,EMP010
2,ORD00155,Cake,2026-05-07 13:49,Online,Card,5.0,19.95,North,24,1,19.95,EMP016
3,ORD00204,NaN,2026-03-25 13:54,Online,E-Wallet,2.0,20.80,South,22,3,62.40,EMP004
4,ORD00187,Fries,2026-05-11 16:58,Online,E-Wallet,2.0,29.60,West,27,1,29.60,EMP025


Understanding the Dataset Shape and Data

Based on the code result, it is expected to contain **1,000 records**. However, there are **10 missing values** in the **Product Name** column and **10 missing values** in the **Rating** column.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   order_id                  1000 non-null   object 
 1   product_name              990 non-null    object 
 2   order_date                1000 non-null   object 
 3   order_type                1000 non-null   object 
 4   payment_method            1000 non-null   object 
 5   rating                    990 non-null    float64
 6   price                     1000 non-null   float64
 7   region                    1000 non-null   object 
 8   preparation_time_minutes  1000 non-null   int64  
 9   quantity                  1000 non-null   int64  
 10  line_total                1000 non-null   float64
 11  employee_id               1000 non-null   object 
dtypes: float64(3), int64(2), object(7)
memory usage: 93.9+ KB


`df["rating"].fillna(df['rating'].median(), inplace = True)` is a code to fill missing value on rating column with median value from rating column

In [ ]:
df["rating"].fillna(df['rating'].median(), inplace = True)

/tmp/ipykernel_5157/2385228026.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["rating"].fillna(df['rating'].median(), inplace = True)


`df = df.dropna()` is a code to drop any rows that contain missing value inside the dataframe. This action taken because filling product name missing value is not possible

In [ ]:
df = df.dropna()

The `df.describe()` function helps users understand the numerical columns in the dataset by providing several descriptive statistics. Based on the results, we noticed that there are incorrect values in the **Price** column. Specifically, a few rows contain a value of **-3** in the **Price** column.

In [ ]:
df.describe()

,rating,price,preparation_time_minutes,quantity,line_total
count,990.000000,990.000000,990.000000,990.000000,990.000000
mean,3.076768,21.301566,23.911111,3.044444,65.720657
std,1.421404,10.865502,12.367403,1.434814,47.031952
min,1.000000,-3.000000,3.000000,1.000000,2.560000
25%,2.000000,11.955000,13.000000,2.000000,27.442500
50%,3.000000,21.030000,24.000000,3.000000,54.070000
75%,4.000000,30.592500,35.000000,4.000000,96.470000
max,5.000000,39.910000,45.000000,5.000000,199.050000


`df['price'].replace(to_replace=-3,value=3, inplace=True)` is a code that replace -3 incorrect value on price column into 3

In [ ]:
df['price'].replace(to_replace=-3,value=3, inplace=True)

/tmp/ipykernel_5157/1617382692.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['price'].replace(to_replace=-3,value=3, inplace=True)


`del df['line_total']` and `df['line_total'] = df['price'] * df['quantity']` are two lines of code. The first line deletes the **line_total** column from the dataset. The second line creates a new **line_total** column and fills it with the result of multiplying the **price** and **quantity** columns.

In [ ]:
del df['line_total']
df['line_total'] = df['price'] * df['quantity']

`df["month"] = df['order_date'].str[5:7]` is a code that make one new column called month and fill that new column with order date data that have been substringed so it only take the month of every date inside the order date column

In [ ]:
df["month"] = df['order_date'].str[5:7]

`df['month'].replace(to_replace='/2',value='02',inplace=True)` this line of code replace value "/2" to "02" on month column

In [ ]:
df['month'].replace(to_replace='/2',value='02',inplace=True)

/tmp/ipykernel_5157/30991257.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['month'].replace(to_replace='/2',value='02',inplace=True)


`df['region'].replace(to_replace='Unknown',value=df['region'].mode()[0],inplace=True)` is a line of code that replace value "unkown" with the mode of region column

In [ ]:
df['region'].replace(to_replace='Unknown',value=df['region'].mode()[0],inplace=True)

/tmp/ipykernel_5157/3854359143.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['region'].replace(to_replace='Unknown',value=df['region'].mode()[0],inplace=True)


#EDA

In [ ]:
df.groupby(['region']).agg(
    average_preparation_time = ('preparation_time_minutes','mean')
).reset_index().sort_values(by='average_preparation_time',ascending=False)

,region,average_preparation_time
4,West,24.586387
1,East,24.153061
2,North,23.897196
3,South,23.809524
0,Central,23.140000


According to the data, all of our stores have a similar average preparation time, ranging from **23 to 24 minutes**.

In [ ]:
df.loc[(df['region'] == 'West')].groupby(["order_type","product_name"]).agg(
    average_preparation_time = ("preparation_time_minutes","mean")
).sort_values(by="average_preparation_time",ascending=False).head(3)

,,average_preparation_time
order_type,product_name,
Dine-In,Burger,37.0
Pickup,Fries,36.0
Online,Burger,34.0


According to the data, at our **West** restaurant, **burgers** (dine-in and online) and **fries** (pickup) take the longest time to prepare on average.

In [ ]:
df.loc[(df['region'] == 'East')].groupby(["order_type","product_name"]).agg(
    average_preparation_time = ("preparation_time_minutes","mean")
).sort_values(by="average_preparation_time",ascending=False).head(3)

average_preparation_time
order_type product_name                          
Takeaway   Cake                              38.0
           Sandwich                          34.5
Online     Burger                            33.0

According to the data, at our **East** restaurant, **Cake** (takeaway), sandwich (takeaway) and **Burger** (online) take the longest time to prepare on average.

In [ ]:
df.loc[(df['region'] == 'North')].groupby(["order_type","product_name"]).agg(
    average_preparation_time = ("preparation_time_minutes","mean")
).sort_values(by="average_preparation_time",ascending=False).head(3)

average_preparation_time
order_type product_name                          
Online     Wrap                         39.666667
           Rice Bowl                    38.333333
Dine-In    Salad                        37.750000

According to the data, at our **North** restaurant, **Wrap** (Online), Rice Bowl (Online) and **Salad** (Dine-in) take the longest time to prepare on average.

In [ ]:
df.loc[(df['region'] == 'South')].groupby(["order_type","product_name"]).agg(
    average_preparation_time = ("preparation_time_minutes","mean")
).sort_values(by="average_preparation_time",ascending=False).head(3)

average_preparation_time
order_type product_name                          
Dine-In    Pasta                        45.000000
Online     Latte                        36.666667
           Smoothie                     36.500000

According to the data, at our **South** restaurant, **Pasta** (Dine-in), **Latte** (Online) and **Smoothie** (Online) take the longest time to prepare on average.

In [ ]:
df.loc[(df['region'] == 'Central')].groupby(["order_type","product_name"]).agg(
    average_preparation_time = ("preparation_time_minutes","mean")
).sort_values(by="average_preparation_time",ascending=False).head(3)

average_preparation_time
order_type product_name                          
Online     Pizza                             34.0
           Smoothie                          33.0
Dine-In    Burger                            31.0

According to the data, at our **South** restaurant, **Pizza** (Online), **Smoothie** (Online) and **Burger** (Dine-in) take the longest time to prepare on average.

In [ ]:
df.groupby(['month']).agg(
    rating_on_avg=('rating','mean'),
).reset_index().sort_values(by='rating_on_avg').head(3)

,month,rating_on_avg
1,02,2.967949
3,04,2.994220
4,05,3.027778


From this result, we understand that second, fourth, fifth months have the lowest rating in this datasets, ranging from 2.9 - 3

In [ ]:
df[df['month'] == '02'].groupby(['region']).agg(
    rating_on_avg=('rating','mean'),
).reset_index().sort_values(by='rating_on_avg').head(3)

,region,rating_on_avg
2,North,2.75
3,South,2.80
4,West,3.00


From this result we know that on second month our restaurant in north, south, and west region have the lowest rating. They range around 2 - 3

In [ ]:
df[df['month'] == '04'].groupby(['region']).agg(
    rating_on_avg=('rating','mean'),
).reset_index().sort_values(by='rating_on_avg').head(3)

,region,rating_on_avg
2,North,2.692308
4,West,2.935484
1,East,2.976744


From this result we know that on forth month our restaurant in north, west, east region have the lowest rating. They range around 2.6 - 2.9

In [ ]:
df[df['month'] == '05'].groupby(['region']).agg(
    rating_on_avg=('rating','mean'),
).reset_index().sort_values(by='rating_on_avg').head(3)

,region,rating_on_avg
2,North,2.731707
3,South,2.970588
0,Central,3.000000


From this result we know that on fifth month our restaurant in north, south, and central region have the lowest rating. They range around 2.7 - 3

In [ ]:
df.groupby(['region']).agg(
    rating_on_avg=('rating','mean'),
).reset_index().sort_values(by='rating_on_avg').head(3)

,region,rating_on_avg
1,East,2.948980
2,North,2.962617
4,West,3.141361


Overall, our restaurant in east, north, and west have the lowest rating from 5 of our restaurant. their rating raging from 2.9 - 3.1

In [ ]:
df[df['region'] == 'East'].groupby('product_name').agg(
    rating_on_avg=('rating','mean')
).reset_index().sort_values(by='rating_on_avg').head(3)

,product_name,rating_on_avg
3,Fries,1.80000
4,Latte,2.47619
2,Coffee,2.62500


In [ ]:
df[df['region'] == 'East']['preparation_time_minutes'].mean()

np.float64(24.153061224489797)

if we focus on our restaurant in east, the lowest rating come from their fries, followed by latte and coffe.

In [ ]:
df[df['region'] == 'North'].groupby('product_name').agg(
    rating_on_avg=('rating','mean')
).reset_index().sort_values(by='rating_on_avg').head(3)

,product_name,rating_on_avg
7,Pizza,2.307692
4,Latte,2.500000
11,Smoothie,2.631579


In [ ]:
df[df['region'] == 'North']['preparation_time_minutes'].mean()

np.float64(23.897196261682243)

And for the second lowest rating is our restaurant in North, the lowest rating come from our Pizza, followed by latte and smoothie.

In [ ]:
df[df['region'] == 'West'].groupby('product_name').agg(
    rating_on_avg=('rating','mean')
).reset_index().sort_values(by='rating_on_avg').head(3)

,product_name,rating_on_avg
5,Noodles,2.636364
2,Coffee,2.727273
8,Rice Bowl,2.833333


In [ ]:
df[df['region'] == 'West']['preparation_time_minutes'].mean()

np.float64(24.586387434554975)

And for the last lowest rating is our restaurant in West, the lowest rating come from our Noodle, followed by Coffe and Rice Bowl.

In [ ]:
df.groupby(['order_type','region']).agg(number_of_order = ("order_id",'count'))

number_of_order
order_type region                  
Dine-In    Central               61
           East                  64
           North                 84
           South                 67
           West                  56
Online     Central               69
           East                  65
           North                 67
           South                 46
           West                  67
Pickup     Central                1
           East                   3
           South                  3
           West                   3
Takeaway   Central               69
           East                  64
           North                 63
           South                 73
           West                  65

According to our data, customers in the **North** region prefer to dine in at our store. Customers in the **West** and **South** regions prefer to place online orders. For takeaway orders, customers prefer the **South** restaurant.

In [ ]:
df[df['region'] == 'Central'].groupby(['region','product_name']).agg(number_of_order = ("order_id",'count')).reset_index().sort_values(by='number_of_order',ascending=False).head(3)

,region,product_name,number_of_order
13,Central,Wrap,22
8,Central,Rice Bowl,21
2,Central,Coffee,18


According to the data, the top-selling items at our **Central** restaurant are **wraps, rice bowls, and coffee**.

In [ ]:
df[df['region'] == 'East'].groupby(['region','product_name']).agg(number_of_order = ("order_id",'count')).reset_index().sort_values(by='number_of_order',ascending=False).head(3)

,region,product_name,number_of_order
4,East,Latte,21
7,East,Pizza,21
11,East,Smoothie,17


According to the data, the top-selling items at our **East** restaurant are **Latte, Pizza, and Smoothie**.


In [ ]:
df[df['region'] == 'South'].groupby(['region','product_name']).agg(number_of_order = ("order_id",'count')).reset_index().sort_values(by='number_of_order',ascending=False).head(3)

,region,product_name,number_of_order
4,South,Latte,21
12,South,Tea,20
9,South,Salad,19


According to the data, the top-selling items at our **South** restaurant are **Latte, Tea, and Salad**.

In [ ]:
df[df['region'] == 'West'].groupby(['region','product_name']).agg(number_of_order = ("order_id",'count')).reset_index().sort_values(by='number_of_order',ascending=False).head(3)

,region,product_name,number_of_order
4,West,Latte,18
10,West,Sandwich,18
3,West,Fries,16


According to the data, the top-selling items at our **East** restaurant are **Latte, Sandwich, and Fries**.

In [ ]:
df[df['region'] == 'North'].groupby(['region','product_name']).agg(number_of_order = ("order_id",'count')).reset_index().sort_values(by='number_of_order',ascending=False).head(3)

,region,product_name,number_of_order
12,North,Tea,22
1,North,Cake,20
11,North,Smoothie,19


According to the data, the top-selling items at our **North** restaurant are **tea, cake, smoothie**.

In [ ]:
df.groupby(['region','payment_method']).agg(number_of_order = ("order_id",'count'))

number_of_order
region  payment_method                 
Central Card                         69
        Cash                         59
        E-Wallet                     72
East    Card                         63
        Cash                         77
        E-Wallet                     56
North   Card                         76
        Cash                         77
        E-Wallet                     61
South   Card                         53
        Cash                         74
        E-Wallet                     62
West    Card                         54
        Cash                         76
        E-Wallet                     61

Based on the data, customers at the **Central** restaurant prefer to pay for their food using **e-wallets (72 orders)**. However, at our other restaurants (**East, North, South, and West**), customers prefer to pay with **cash (around 74 - 76 orders)**.

In [ ]:
df.groupby(['month']).agg(revenue = ('line_total','sum')).reset_index().sort_values(by='revenue',ascending=False)

,month,revenue
3,04,11934.48
5,06,11545.84
4,05,11522.91
0,01,10203.98
1,02,9969.70
2,03,9273.65


According to the data, the restaurant generated the highest revenue in **April (11k), June(11k), and May(11k)**, while the lowest revenue was recorded in **March (9k)**.

In [ ]:
df.groupby(['month']).agg(order = ('order_id','count')).reset_index().sort_values(by='order',ascending=False).head(2)

,month,order
4,05,180
3,04,173


According to the data, May and April had the highest number of orders compared to the other months, with 180 and 173 orders, respectively.

In [ ]:
df[df['month'] == '05'].groupby(['product_name','order_type']).agg(order = ('order_id','count')).reset_index().sort_values(by='order',ascending=False).head(3)

,product_name,order_type,order
12,Latte,Dine-In,11
13,Latte,Online,10
37,Tea,Dine-In,8


According to the data, on May latte (dine-in and online) has the highest order number (11 order) followed by tea (dine in) with 8 order in that period

In [ ]:
df[df['month'] == '04'].groupby(['product_name','order_type']).agg(order = ('order_id','count')).reset_index().sort_values(by='order',ascending=False).head(3)

,product_name,order_type,order
6,Coffee,Dine-In,8
36,Tea,Online,8
1,Burger,Online,7


According to the data, on april coffee (dine-in) is the number one item that get ordered (8 order), followed by tea through online order (8 order), and burger (online) with 7 order during that month

In [ ]:
df.to_csv('fnb.csv')

In [ ]:
df.head(4)

,order_id,product_name,order_date,order_type,payment_method,rating,price,region,preparation_time_minutes,quantity,employee_id,line_total,month
0,ORD00334,Rice Bowl,2026-02-27 08:33,Takeaway,Card,5.0,36.76,South,33,4,EMP026,147.04,02
1,ORD00284,Noodles,2026-03-03 14:35,Takeaway,Cash,3.0,14.31,East,32,3,EMP010,42.93,03
2,ORD00155,Cake,2026-05-07 13:49,Online,Card,5.0,19.95,North,24,1,EMP016,19.95,05
4,ORD00187,Fries,2026-05-11 16:58,Online,E-Wallet,2.0,29.60,West,27,1,EMP025,29.60,05
